$ $
#  Ejercicio 2.12 - Aproximacion de $\pi$  por series infinitas

# --------------------------------------------------------------------------------------------------------------------

### sabemos que tan (pi / 4) = 1. Ahora podemos usar la función de tangente inversa, arctan (x), para calcular arctan (1) = pi / 4. Y afortunadamente, tenemos una fórmula simple y fácil para arctan (x). Este método también se conoce como la serie Gregory-Leibniz o la serie Madhava-Gregory, que lleva el nombre de Gottfried Leibniz (1646), James Gregory (1638) y Madhava de Sangamagrama (1340).

$ $
# $arctan(x) = x - \frac{x^{3}}{3} + \frac{x^{5}}{5} - \frac{x^{7}}{7} ... =  \sum_{n=0}^{\infty }\frac{(-1)^{n}x^{2n+1}}{2n+1} $

### Podemos establecer x = 1 y dejar que se ejecute el siguiente código.

In [2]:
import math
import time

def gregory_leibinz():
    pi = 0
    start = time.time()
    for x in range(0, 10000000):#1000000000
        numerador = math.pow(-1, x) * math.pow(1, 2*x+1)
        denominador = 2*x + 1
        pi += numerador/denominador
    end = time.time()
    pi = pi * 4
    print('Duracion de compilacion: ', end - start)
    print("Valor estimado: " + str(pi))
    print("Valor actual:   " + str(math.pi))

def machin():
    num_iteraciones = 6
    arc_tan_1_239 = arc_tan(1/239, num_iteraciones)
    arc_tan_1_5 = arc_tan(1/5, num_iteraciones)
    pi = 4 * (4*arc_tan_1_5 - arc_tan_1_239)
    print("Valor estimado: " + str(pi))
    print("Valor actual: " + str(math.pi))


def arc_tan(x, iteraciones):
    arc_tan_x = 0
    for z in range(0, iteraciones):
        numerador = math.pow(-1, z) * math.pow(x, 2 * z + 1)
        denominador = 2 * z + 1
        arc_tan_x += numerador / denominador
    return arc_tan_x

if __name__ == "__main__":
    print('--------leibniz--------')
    gregory_leibinz()
    print('---------machin--------')
    machin()

--------leibniz--------
Duracion de compilacion:  5.622813940048218
Valor estimado: 3.1415925535897915
Valor actual:   3.141592653589793
---------machin--------
Valor estimado: 3.1415926526153086
Valor actual: 3.141592653589793


### millones de iteraciones, obtenemos un dígito más que con el método de integración.

| numero de iteraciones | valor estimado (incluye cifras significativas) |
|-----------------------|------------------------------------------------|
| 10.000.000 (10m)      | 3.1415925                                      |
| 100.000.000 (100m)    | 3.14159264                                     |
| 1000.000.000 (1b)     | 3.141592652                                    |

Podemos continuar con este método con algunas mejoras y obtener resultados drásticamente mejores. Una mejora proviene de John Machin (1706), quien desarrolló la siguiente fórmula. 

$ $
# $\frac{\pi}{4} = 4\cdot arctan\frac{1}{5} - arctan\frac{1}{239}$

Los resultados están a continuación. Asegúrese de anotar el número de iteraciones. Con 10 iteraciones, estamos estimando tan bien como arriba. El primer método de requirió mil millones de iteraciones para hacer lo que hizo la fórmula de Machin en 3.

| numero de iteraciones | valor estimado (incluye cifras significativas) |
|-----------------------|------------------------------------------------|
| 3                     | 3.1416                                         |
| 6                     | 3.141592652                                    |
| 10                    | 3.141592654                                    |